source/particle/generator/probability_density_function.cc

```diff
@@ -0,0 +1,295 @@
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file doc/COPYING.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/particle/generator/probability_density_function.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/base/geometry_info.h>
+
+#include <boost/random.hpp>
+
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Generator
+    {
+      template <int dim>
+      ProbabilityDensityFunction<dim>::ProbabilityDensityFunction()
+        :
+        random_number_generator(5432),
+        n_tracers()
+      {}
+
+      template <int dim>
+      std::multimap<types::LevelInd, Particle<dim> >
+      ProbabilityDensityFunction<dim>::generate_particles()
+      {
+        // Get the local accumulated probabilities for every cell
+        const std::vector<double> accumulated_cell_weights = local_accumulated_cell_weights();
+
+        // Sum the local integrals over all nodes
+        double local_function_integral = accumulated_cell_weights.back();
+        const double global_function_integral = Utilities::MPI::sum (local_function_integral,
+                                                                     this->get_mpi_communicator());
+
+        AssertThrow(global_function_integral > std::numeric_limits<double>::min(),
+                    ExcMessage("The integral of the user prescribed probability density function over the domain equals zero, "
+                               "ASPECT has no way to determine the cell of generated tracers. Please ensure that the provided "
+                               "function is positive in at least a part of the domain, also check the syntax of the function."));
+
+        // Determine the starting weight of this process, which is the sum of
+        // the weights of all processes with a lower rank
+        double start_weight = 0.0;
+        MPI_Scan(&local_function_integral, &start_weight, 1, MPI_DOUBLE, MPI_SUM, this->get_mpi_communicator());
+        start_weight -= local_function_integral;
+
+        // Build the map between integrated probability density and cell
+        std::map<double, types::LevelInd> cell_weights;
+        typename DoFHandler<dim>::active_cell_iterator cell = this->get_dof_handler().begin_active(),
+                                                       endc = this->get_dof_handler().end();
+        for (unsigned int i = 0; cell!=endc; ++cell)
+          if (cell->is_locally_owned())
+            {
+              // adjust the cell_weights according to the local starting weight
+              cell_weights.insert(std::make_pair(accumulated_cell_weights[i] + start_weight, types::LevelInd(cell->level(),cell->index())));
+              ++i;
```

Comment:
I documented the size, but still only loop over all locally owned active cells, because then the global integral is only the MPI sum of all last entries (if I had ghost cells in there, I would need to remove duplications, and everything would get awfully complicated).